### PR TITLE
Fix bug to check parent file when meta is checked

### LINKED ISSUE #913: Committing new file by selecting its meta file does not commit the file itself
### Prerequisites
Tracing enabled, log file attached (but not much to see in it).
[github-unity.log](https://github.com/github-for-unity/Unity/files/2398910/github-unity.log)

### Description

When a new file is created, and the file is marked to be committed by selecting its meta file, the file itself will not be committed, even though the tree view shows it as also selected.

### Steps to Reproduce

1. Initialize GitHub for Unity, and a git repository
2. Create any new file in the project.
3. In the GitHub for Unity window click on the meta file of the newly created file. Notice that the newly created file is also marked as selected, as the meta file is a dependency of it.
4. Add commit message, and click on commit.

**Expected behavior:** All the files from the GitHub window that have their checkbox ticked in the tree view are (staged and) committed.

**Actual behavior:** Only the meta file is committed, without the file for which it was created.

**Reproduces how often:** 100% (Always)

### Additional Information
See attached "empty" Unity project, already in inconsistent state. Follow instructions to reproduce issue.
[GitCommitIssue.zip](https://github.com/github-for-unity/Unity/files/2398940/GitCommitIssue.zip)

Unity version is 2018.2.0f2, GitHub for Unity version is latest on Asset Store as of 2018-09-20.

It is probably a UI issue, inconsistency between what the UI logic shows and the list of files to be committed.

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes #913 

As a result of an item having it's check state changed in our tree lists, the check state of it's parents or children might need to be changed.

When performing this action from `TreeBase<TNode, TData>` we need to be sure to call `AddCheckedNode(...)` or `RemoveCheckedNode(...)` when those items are not folders. Those functions are overloaded and allow derived classes to set some additional state.

In order to do this cleanly, all nodes check states are now being changed through a central overloaded method.

https://github.com/github-for-unity/Unity/blob/d8d92c990278df1306754c22ed90de05193686ae/src/GitHub.Api/UI/TreeBase.cs#L303-L328

### CODE REVIEW

File: src/GitHub.Api/UI/TreeBase.cs [hunk 1]
```diff
@@ -332,8 +300,34 @@ private List<TNode> GetLeafNodes(TNode node, int idx)
...
+
+            var wasChecked = node.CheckState == CheckState.Checked;
+
+            node.CheckState = setCheckState;
+
+            if (!node.IsFolder)
+            {
+                if (isChecked && !wasChecked)
+                {
+                    AddCheckedNode(node);
```
Comment (queerviolet) [inline src/GitHub.Api/UI/TreeBase.cs]: Out of curiosity, what do `AddCheckedNode` and `RemoveCheckedNode` do? Are they callbacks—like, "this node is being added/removed to/from the set of checked nodes?"

File: src/GitHub.Api/UI/TreeBase.cs [hunk 2]
```diff
@@ -332,8 +300,34 @@ private List<TNode> GetLeafNodes(TNode node, int idx)
...
+                {
+                    RemoveCheckedNode(node);
+                }
+            }
+        }
+
+        private void ToggleParentFolderAndContainersChecked(int idx, TNode node, bool isChecked)
         {
             while (true)
             {
```
Comment (queerviolet) [inline src/GitHub.Api/UI/TreeBase.cs]: wow does it ever seem like there should be a simpler way to track a bunch of checkboxes.

(I'm not saying there is. Just that it seems like there ought to be.)

File: src/GitHub.Api/UI/TreeBase.cs [hunk 3]
```diff
@@ -384,14 +378,15 @@ private void ToggleParentFoldersChecked(int idx, TNode node, bool isChecked)
...
-                    if (siblingsInSameState)
-                    {
-                        parentNode.CheckState = isChecked ? CheckState.Checked : CheckState.Empty;
-                    }
-                    else
-                    {
-                        parentNode.CheckState = CheckState.Mixed;
-                    }
+
+                    var parentNodeState =
```
Comment (queerviolet) [inline src/GitHub.Api/UI/TreeBase.cs]: I <3 nested ternaries.

File: src/tests/UnitTests/UI/TreeBaseTests.cs [hunk 4]
```diff
@@ -606,6 +611,74 @@ public void ShouldPopulateTreeWithSingleEntryWithMetaInPath()
                 }
             });
         }
+
+        [Test]
+        public void ShouldCheckParentOfMetaFile()
```
Comment (queerviolet) [inline src/tests/UnitTests/UI/TreeBaseTests.cs]: I know these are a pain to set up, but could we maybe do a test where we have
  - [X] Root
    - [x] Parent
      - [X] A
      - [X] B
      - [x] C

And then we uncheck `C` and see `Parent` and `Root` both go grey?

**Comment by StanleyGoldman** [comment]
@queerviolet 🏆. That was a good call. You made me find two bugs.

File: src/GitHub.Api/UI/TreeBase.cs (hunk 2 above)
Comment (queerviolet) [reply src/GitHub.Api/UI/TreeBase.cs]: Yeah, I figured.

File: src/tests/UnitTests/UI/TreeBaseTests.cs [hunk 5]
```diff
@@ -606,6 +611,196 @@ public void ShouldPopulateTreeWithSingleEntryWithMetaInPath()
...
+            testTree.ToggleNodeChecked(5, cNode);
+
+            // Checked c.txt, everything checked
+
+            Assert.AreEqual(CheckState.Checked, rootNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, parentNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, aNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, bNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, cNode.CheckState);
+
```
Comment (queerviolet) [inline src/tests/UnitTests/UI/TreeBaseTests.cs]: Can we do one more where we uncheck `a.txt`, `b.txt`, and `c.txt`, and verify that the parent unchecks entirely?

Comment (queerviolet) [reply src/tests/UnitTests/UI/TreeBaseTests.cs]: Actually, the whole tree should uncheck at that point, no? That feels like a really appropriate final assertion. ;)

**Comment by queerviolet** [comment]
Yay! I have one more test request, then this looks good to go. 🙏

File: src/tests/UnitTests/UI/TreeBaseTests.cs (hunk 5 above)
Comment (StanleyGoldman) [reply src/tests/UnitTests/UI/TreeBaseTests.cs]: No problem

**Pull request merged**